feat: add a `detach` mode to the CLI

jeffsmale90 opened the pull request:
Do you need to detach? Do you need to do it _now_?

It can be annoying needing an extra terminal open just to run Ganache. This becomes super annoying when you want to have different instances and flavors of Ganache running at the same time.

Enter: Detach mode.

Start Ganache with the `--detach` flag, and it'll return to the console as soon as Ganache is ready to receive requests (writing the name of the instance to stdout).

`ganache instances list` will show you all of the instances of Ganache running in Detach mode, and you can stop them with `ganache instances stop <name>`.

```
┌───────┬─────────────────────────┬──────────┬─────────┬────────────────┬────────┐
│   PID │ Name                    │ Flavor   │ Version │ Host           │ Uptime │
├───────┼─────────────────────────┼──────────┼─────────┼────────────────┼────────┤
│ 12182 │ salted_caramel_ganache  │ ethereum │ 7.6.0   │ 127.0.0.1:8545 │ 2m 46s │
├───────┼─────────────────────────┼──────────┼─────────┼────────────────┼────────┤
│ 53184 │ candied_caramel_truffle │ ethereum │ 7.6.0   │ 127.0.0.1:8545 │ 2m 56s │
└───────┴─────────────────────────┴──────────┴─────────┴────────────────┴────────┘
```

With the following command, you can start Ganache, run your tests, and stop Ganache when you are finished.

```Bash
GANACHE=$(ganache --detach) && npm run test; ganache instances stop $GANACHE
```

Or if you are running PowerShell on Windows, you can do:

```PowerShell
$GANACHE=ganache --detach; npm run test; ganache instances stop $GANACHE
```

Fixes: #1001

jeffsmale90 committed:
feat: allow users to specify a  option, which causes Ganache to start up as a background process

jeffsmale90 committed:
Stream stderr from the child to the parent

jeffsmale90 committed:
Tidy up of --detach argument parsing

jeffsmale90 commented on src/packages/cli/src/args.ts:
```diff
@@ -219,6 +221,12 @@ export default function (version: string, isDocker: boolean) {
 
             return true;
+          })
+          .option("cli.detach", {
```
I'm not sure on this name-spacing:
➕  It doesn't necessitate modification to the logic breaking out the namespaces
➕  With a default value, we can access `args.cli.detach` without having to check whether `cli === undefined`
➖  It adds an unnecessary namespace, which we are locked into maintaining
➖  It pollutes the help text

I'm also unsure about placement in the help text - presently it's at the bottom with "general" options, where I'm not sure if fits, but I'm also not keen to put in a separate "Cli" section, with _just_ detach.

Keen to hear people's opinions.

```
Server:
      --server.ws                        Enable a websocket server.                            [boolean] [default: true]
      --server.wsBinary                  Whether or not websockets should response with binary data (ArrayBuffers) or
                                         strings.                     [choices: "true", "false", "auto"] [default: auto]
      --server.rpcEndpoint               Defines the endpoint route the HTTP and WebSocket servers will listen on.
                                                                         [default: "/" (Ethereum), "/rpc/v0" (Filecoin)]
      --server.chunkSize                 For memory and performance reasons ganache may respond with chunked
                                         transfer-encoding over HTTP and fragmented send over WebSockets. This option
                                         allows you to control the approximate size of each chunk.
                                                                                             [number] [default: 1048576]
  -h, --server.host, --host, --hostname  Hostname to listen on.
                                         deprecated aliases: --host, --hostname
                                                                                         [string] [default: "127.0.0.1"]
  -p, --server.port, --port              Port to listen on.
                                         deprecated aliases: --port
                                                                                                [number] [default: 8545]

Options:
  -?, --help                  Show help                                                                        [boolean]
      --version               Show version number                                                              [boolean]
  -D, --cli.detach, --detach  Run Ganache in detached mode.                               [boolean] [default: (default)]
  ```

gnidan replied:
What if the namespace were called `process`? That seems more widely useful for future expansion, and also more precise a term

jeffsmale90 replied:
We discussed this today, and, although process might be a better term than cli, it's confusing as it's both a verb and a noun.

We agreed that actually _having_ a namespace was a workaround to the implementation, so we changed the implementation to allow `--detach` :)

This is locking in the name `--detach`, but if we decide in the future to namespace, we can keep that as an alias, so no breaking changes :)

jeffsmale90 committed:
Improve detach configuration

jeffsmale90 committed:
Add tests for parsing detach arguments. Change default value to false (but don't show that in the help, because it's ugly)

davidmurdoch commented on src/packages/cli/src/args.ts:
```diff
@@ -222,11 +222,10 @@ export default function (version: string, isDocker: boolean) {
-          .option("cli.detach", {
-            description: "Run Ganache in detached mode.",
+          .option("detach", {
+            description: "Run Ganache in detached (daemon) mode",
```
Can you add some additional documentation here? Something about what to do to kill it (or attach) to it.

davidmurdoch commented on src/packages/cli/src/cli.ts:
```diff
@@ -39,108 +40,158 @@ const isDocker =
+    // so let the user know (even if the exit code is 0).
+    console.error(`The child process exited with exit code ${code}`);
+    process.exit(code);
+  });
```
It'd be nice if this logic was in its own file, especially since the diff in the PR Review on this one is an absolute mess.

davidmurdoch commented on src/packages/cli/src/cli.ts:
```diff
@@ -39,108 +40,158 @@ const isDocker =
+    console.error(
+      `An error occurred starting Ganache in detached mode: ${err}`
+    );
+    process.exit(1);
```
The `process.exit()` forcefully closes the process immediately without letting running timers or promises finish up. While it is unlikely there are any async things happening here, that doesn't mean things won't change in the future.

The reason we use `process.exit(1)` in the failed `server` initialization case elsewhere in this file is because failing to instantiate the `server` could mean there are promises, websockets, etc, that were started in the background, but something has gone horribly wrong and we now have no way to clean up after ourselves (since we don't get a `server` instance we can call `close()` on). We forcibly close the whole process in an attempt to redeem any poor ~life decisions~ programming chooses we've made. Admittedly not the best solution to the problem, but that is the story as to why I justified the use of `process.exit(1)` there.

```suggestion
    process.exitCode = 1;
```

davidmurdoch commented on src/packages/cli/src/cli.ts:
```diff
@@ -39,108 +40,158 @@ const isDocker =
+    // If the child process exits before the parent, something has gone wrong,
+    // so let the user know (even if the exit code is 0).
+    console.error(`The child process exited with exit code ${code}`);
+    process.exit(code);
```
Probably should not use `exit()` here, either.

```suggestion
    process.exitCode = code;
```

davidmurdoch commented on src/packages/cli/tests/args.test.ts:
```diff
@@ -0,0 +1,56 @@
+import assert from "assert";
+import args from "../src/args";
+
+describe("args", () => {
```
YAY! CLI TESTS!

You should remove the `test/index.test.ts` file now, as it was a placeholder for tests.

Since these tests only test that the `detach` option is set, and not if it works, let's find a decent `describe` title or `it` titles that makes that clear.

davidmurdoch commented on src/packages/cli/tests/args.test.ts:
```diff
@@ -0,0 +1,56 @@
+      "--detach=false",
+      "--D=false",
+      "--😈=false"
+    ];
```
I didn't know about `--no-*`. Neat!

davidmurdoch commented on src/packages/cli/tests/args.test.ts:
```diff
@@ -0,0 +1,56 @@
+    const versionString = "Version string";
+    const isDocker = false;
+
+    const detachModeArgs = ["--detach", "--D", "--😈"];
```
Should we add boolean as well here? Like `--detach=true`? Or do you think it's overkill?

jeffsmale90 committed:
Make parent aware of detached instances, and allow stopping them via cli

jeffsmale90 committed:
improve args tests, remove placeholder

jeffsmale90 committed:
Implement , and allow different actions to be executed via the CLI

jeffsmale90 committed:
Clean up orphaned instance process files. uses a terrible hack to fetch pid list.

jeffsmale90 committed:
Implement `instances` command with subcommands. Clean up orphaned instance files if the process no longer exists.

davidmurdoch commented on src/packages/cli/src/process-name.ts:
```diff
@@ -0,0 +1,83 @@
+  "spiced",
+  "sticky"
+];
+const flavours = [
```
```suggestion
const flavors = [
```

:-p

jeffsmale90 committed:
Tidy up args types and parsing

jeffsmale90 committed:
Refactor, add flavor to instance data, and to list output. Change instance data file to JSON format.

jeffsmale90 committed:
Fix createFlatChildArgs import in args.test.ts

jeffsmale90 committed:
Merge branch 'develop' into feat/detach-mode

jeffsmale90 committed:
Update package-lock.json files - required for new deps

jeffsmale90 committed:
Fix module resolution

jeffsmale90 committed:
Tidy ups and refactors

jeffsmale90 committed:
Tidy up arg parsing

jeffsmale90 committed:
Revert package-lock.json files

jeffsmale90 committed:
Fix miss-typed action name

tenthirtyone commented on src/packages/cli/package.json:
```diff
@@ -62,6 +62,8 @@
   },
   "dependencies": {
+    "table": "6.8.0",
+    "ps-list": "7.2.0",
```
I saw the manual verification on windows, this package has lots of downloads, I feel confident it will not but - are we sure that WIndows or some virus thing is not going to flag the packaged binaries as malware or get in the way of the all the snowflake UAC configurations that exist? (If Windows still has all that stuff).

JS executing an `.exe` feels like 
![image](https://user-images.githubusercontent.com/6285466/191357632-f173cca4-20bb-4535-9e69-cbee5ffe1099.png)

Also, will the compiled exe's run in the browser or affect browser compatibility/security?

jeffsmale90 replied:
Initially I removed the `ps-list` dep, but added it back in because we decided we probably do need to concern ourselves with the scenario where a different process has snatched the PID. We check that by recording the `process.cmd` when we start it up, and ensuring that it matches when we iterate the detached instances.

![image](https://user-images.githubusercontent.com/6363749/191770426-439c0b0f-7516-49e7-b047-3c2c77a01449.png)

jeffsmale90 replied:
As for the concerns about executing an .exe - I'm not really sure. Is it wildly different to executing any other binaries (such as the websockets??)

Maybe @davidmurdoch has an opinion? Otherwise I can research.

davidmurdoch replied:
Why are you using an old version?

jeffsmale90 replied:
I'm using an old version as the next version removes commonjs support. I've created a fork that:
- Supports CJS
- Uses the latest non-Windows implementation
- Uses the native `task-list` Windows utility (rather than `fast-list`)

Which can be found here: https://github.com/trufflesuite/ps-list

Once the pulls on that are merged, I'll publish, and add the dependency here.

tenthirtyone commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,198 @@
+    if (pids.some(p => p === parseInt(pid))) {
+      const filepath = path.join(dataPath, pid);
+      const content = readFileSync(filepath).toString("utf8");
+      const instance = JSON.parse(content) as DetachedInstance;
```
I'm dated but `JSON.parse` throws. Not sure of "Types are Enough" prevents needing a defensive try/catch

jeffsmale90 replied:
I don't think it's overly likely, but it would be super annoying if `ganache instances` stopped working because there was a file with invalid JSON.

I've put a try/catch around the entire parsing of the file, and throw the instance out if it fails.

tenthirtyone commented on src/packages/cli/src/process-name.ts:
```diff
@@ -0,0 +1,85 @@
+ * generated from an adjective, a flavor and a type of desert, in the form of
+ * `<adjective>_<flavor>_<type>`, eg., `salted_caramel_ganache`.
+ */
+export default function createInstanceName() {
```
`awful` needs to be a flavor so we can have a possibility of an `awful_waffle`

![image](https://user-images.githubusercontent.com/6285466/191349286-bac6f3e8-a772-4293-ab01-8a90fddc1448.png)

jeffsmale90 replied:
Maybe we need an Align meeting to brainstorm appropriate instance name parts :P

tenthirtyone commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,198 @@
+  flavor: FlavorName;
+};
+
+const dataPath = envPaths(`ganache`).data;
```
Suggestion: `Ganache/detach` or `Ganache/instances` or something else

Ref: https://github.com/trufflesuite/ganache/blob/274d5521a66a13e9dded95e1109383f70104b6de/src/chains/ethereum/ethereum/src/forking/persistent-cache/persistent-cache.ts#L106

tenthirtyone reviewed (commented):
I think we should assign random ports or detect if a detached instance is running on the default ganache ports. 

People assume, forget, don't read output (like me). If I leave a detached instance on default ports running for a week and then forget about it and try to run ganache it will bomb out - it will tell me the default port is in use in the error message, as if I had started ganache in separate terminals. I'll look at my terminals, go `wtf`, and hopefully check detached instances after confirming ganache is not running in the fg somewhere. 
 
If we assign rando ports for detached instances - how do we alert users? I think it's a bigger question for a separate issue. 

However, we should check if there are detached instances and if a user is starting ganache with a default port (or any port that could possibly collide) from cli without detach, alert them a detached instance is running on that port already instead of re-using the existing error message.

tenthirtyone commented on src/packages/cli/package.json:
```diff
@@ -62,6 +62,8 @@
   },
   "dependencies": {
+    "table": "6.8.0",
+    "ps-list": "7.2.0",
```
Also, follow up to this. I see you are `mapping`/`some` over the pids. Since you are storing a pidfile, you may be able to just 

`process.kill(PID, 0)` without any additional packages. 

Passing a `0` to kill performs all the work of kill, but never sends the signal. If the pid does not exist it does throw but so does the way `ps-list` does it. If the pid exists it just returns true.

This is the implementation: https://github.com/nisaacson/is-running/blob/master/index.js

tenthirtyone commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,198 @@
+      const instance = JSON.parse(content) as DetachedInstance;
+      instances.push(instance);
+    } else {
+      removeDetachedInstanceFile(parseInt(pid));
```
When will this run?

I tried `kill` and `reboot` and so far as I can tell the pid file [is removed on exit.](https://github.com/trufflesuite/ganache/pull/3568/files#diff-d5cc68b0e026fee131b90a36b9f5bb25b719c702a557cb904ecde02dda59f78cR29) How can I create a state where this fires?

tenthirtyone replied:
Scratch that, did a quick power outage plug pull. Nice coverage.

tenthirtyone reviewed (changes requested):
Also, is there a way to squelch or manage the node inspector in dev mode with a detached process? The ports collide,

jeffsmale90 committed:
Remove dependency on ps-list, move data files to Ganache/instances, improve detach help

jeffsmale90 committed:
Remove kebab-case names when expanding arguments. Add tests for expandArgs()

jeffsmale90 committed:
Reintroduce ps-list, only list instances where the process co
mmand matches (otherwise assume the process has been killed)

jeffsmale90 committed:
Move createFlatChildArgs() to detach.ts, add version to list, ensure process.cmd matches.

jeffsmale90 committed:
Create instances directory recursively, because now it's

MicaiahReid commented on src/packages/cli/src/args.ts:
```diff
@@ -1,16 +1,18 @@
 import { TruffleColors } from "@ganache/colors";
+
```
```suggestion
```

jeffsmale90 replied:
I added a section in the README (directly _after_ the startup options) because I didn't feel it fit especially well with the existing options, mostly because of this line:

```The following options are listed for command line use, but can also be used when running Ganache programatically in your project.```

Please let me know what you think of how I've done this, and my wording.

MicaiahReid commented on src/packages/cli/src/args.ts:
```diff
@@ -219,32 +228,116 @@ export default function (version: string, isDocker: boolean) {
+            description: highlight(
+              "Run Ganache in detached (daemon) mode." +
+                EOL +
+                "See `ganache instances --help` for information omn managing detached instances."
```
```suggestion
                "See `ganache instances --help` for information on managing detached instances."
```

MicaiahReid commented on src/packages/cli/src/cli.ts:
```diff
@@ -28,6 +50,20 @@ const logAndForceExit = (messages: any[], exitCode = 0) => {
 
+const formatDuration = (ms: number) => {
+  ms = Math.abs(ms);
+  const time = {
```
This is a neat way to format a duration. Well done!

jeffsmale90 replied:
I'm sorry to say, this wasn't my genius, but inspired/adapted/copied from: https://www.30secondsofcode.org/js/s/format-duration

MicaiahReid commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,256 @@
+}
+
+/**
+ * Flattens parsed, and namespaced args into an array of arguments to be passed
```
```suggestion
 * Flattens parsed and namespaced args into an array of arguments to be passed
```

MicaiahReid commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,256 @@
+/**
+ * Fetch all instance of Ganache running in detached mode. Cleans up any
+ * instance files for processes that are no longer running.
+ * @returns Promise<DetachedInstance[]> resolves with an array of instances
```
```suggestion
 * @returns {Promise<DetachedInstance[]>} resolves with an array of instances
```

MicaiahReid commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,256 @@
+    const filename = files[i];
+    const pid = parseInt(filename);
+
+    const processes = await psList();
```
Can this be called once outside of the loop?

jeffsmale90 replied:
Absolutely it can!

MicaiahReid commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,256 @@
+  const instance = await findDetachedInstanceByName(instanceName);
+  if (instance !== undefined) {
+    try {
+      process.kill(instance.pid, "SIGTERM");
```
Is it possible to `removeDetachedInstanceFile` here after we kill it rather than with the `process.on("exit"...` in the cli file? It would be nice to have the startup and shutdown/cleanup centralized in this one file.

jeffsmale90 replied:
The reason that I do it here, is that we will clean up in the case that the process exits other than via `ganache instances stop`. However we also clean up any instance files when we attempt to access the list of instances.

Initially I preferred this, as we are cleaning up as early as possible, but after looking into it, I agree it's cleaner as you proposed.

MicaiahReid reviewed (changes requested):
Partial review!

I think we should document this feature on our README with the other CLI options. Can you add that as well?

jeffsmale90 committed:
Tidy ups from code review, add section on Detached mode to README.

MicaiahReid commented on src/packages/ganache/README.md:
````diff
@@ -470,7 +470,43 @@ Server:
+```
+Commands:
+  cli.ts instances list         List instances running in detached mode
+  cli.ts instances stop <name>  Stop the instance specified by <name>
````
```suggestion
  ganache instances list         List instances running in detached mode
  ganache instances stop <name>  Stop the instance specified by <name>
```

jeffsmale90 committed:
Fix incorrect command name in README.md

Co-authored-by: Micaiah Reid <micaiahreid@gmail.com>

jeffsmale90 committed:
Merge branch 'develop' into feat/detach-mode

jeffsmale90 committed:
Update to use @trufflesuite/ps-list instead of ps-list

jeffsmale90 committed:
If no instances are found, using , print a friendly message instead of an empty table

jeffsmale90 committed:
Revert ganache package version to match develop

jeffsmale90 committed:
Merge branch 'develop' into feat/detach-mode

jeffsmale90 committed:
Bump @trufflesuite/ps-list to include support for node 12

jeffsmale90 committed:
upgrade @trufflesuite/ps-list to support nodejs < 12.17.0

jeffsmale90 committed:
Merge branch 'develop' into feat/detach-mode

tenthirtyone commented on src/packages/cli/src/cli.ts:
```diff
@@ -8,6 +8,17 @@ import initializeEthereum from "./initialize/ethereum";
+  getDetachedInstances
+} from "./detach";
+import { TruffleColors } from "@ganache/colors";
+import { table } from "table";
```
I may snag this for version check stuff. 

Tangential to this pr, `table` seems extensive and well vetted but it is pretty big after install for what it does. How did you like it and are there other places you think we can use it?

jeffsmale90 replied:
Yeah - it's good enough! I have used `table` in some tools I've written in the past, but not for anything complex.

You triggered me to look back into this, and have switched out `table` for `cli-table`, as that's an existing dependency via `marked-terminal`. The API is almost the same.

I've had to modify the styling, because by default the headers are red 🤮, but it looks mostly indistinguishable now.

tenthirtyone commented on src/packages/ganache/README.md:
````diff
@@ -476,7 +476,43 @@ Server:
+  -D, --detach, --😈                     Run Ganache in detached (daemon) mode.                       [boolean]
+```
+
+This will start Ganache as a background process, and return to the console as soon as Ganache has started and ready to
````
```suggestion
This will start Ganache as a background process, and return to the console as soon as Ganache has started and is ready to
```

tenthirtyone commented on src/packages/ganache/README.md:
````diff
@@ -476,7 +476,43 @@ Server:
+the `ganache instances` command with the following arguments:
+
+```
+Commands:
````
This is cool. 

Is it possible/worth an alias from `ganache instances start [params]` to `ganache [params] --detach` ?

jeffsmale90 replied:
I think that's a great idea. I found myself typing `ganache instances...` when trying to start a new instance, so this would definitely be an improvement.

I made https://github.com/trufflesuite/ganache/issues/3903 to cover that.

tenthirtyone commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,259 @@
+
+  const instanceFilename = `${dataPath}/${instance.pid}`;
+
+  writeFileSync(instanceFilename, JSON.stringify(instance));
```
I believe this will throw in the event that the filesystem/file is readonly or otherwise restricted in access.

jeffsmale90 replied:
Definitely not ideal! I've tested a couple of scenarios here: #3904, and although some of the cases are sub-optimal, I don't think any are critical.

Most importantly:
- If a running instance's file cannot be _read_, the instance is terminated
- if a running instance's file cannot be _deleted_, it is just ignored

tenthirtyone replied:
Cool Cool, 

Yeah the only reason it is in my head is because of my recent NTFS issues flipping between OS's. I keep my `/home` on a separate partition. Sometimes symlinking can be involved and things just aren't available if the drive or partition has issues. 

Once the FS is involved I get a bit hand wavey using `try catch`

tenthirtyone reviewed (changes requested):
First Pass

tenthirtyone reviewed (approved):
👍 none of that should block merging

jeffsmale90 committed:
Improve wording of `--detach` description in README.md

Co-authored-by: Alex Sherbuck <AlexSherbuck@gmail.com>

jeffsmale90 committed:
Switched out  for  (existing dependency via marked-terminal)

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,259 @@
+): Promise<DetachedInstance> {
+  const flavor = args.flavor;
+  const childArgs = createFlatChildArgs(args);
+  const child = fork(module, childArgs, {
```
I don't think this will work in some cases, like `ganache --fork`, or a contrived but probably possible with some combination of arguments: `ganache --namespace.someOption=' <- quoted spaces -> '`. I think we need to pass the original `argv` args, sans `--detach` to the child process.

jeffsmale90 replied:
Fixed

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,259 @@
+    stdio: ["ignore", "ignore", "pipe", "ipc"],
+    detached: true
+  });
+  const cmd = (await psList()).find(p => p.pid === child.pid).cmd;
```
Why do we need to find the cmd when we are the ones that started the command? It'd be nice if we didn't have to `pslist` to get it.

Though `getDetachedInstances` already calls `pslist`, so maybe we could cache the result and it wouldn't matter.

jeffsmale90 replied:
I've changed this to construct the `cmd` property. This works on macOS, and Windows. [ps-list attempts to handle variations of the `ps` implementation for different distros](https://github.com/trufflesuite/ps-list/blob/main/index.js).

I will test this on a wider range of Linux distros - but worst case, the `cmd` property will mismatch, and the process will be left running, but will not appear in `ganache instances list`. Ganache outputs a warning `Process information doesn't match; ${instanceName} has been removed. (PID: ${instance.pid}).`.

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,259 @@
+  // event is emitted) will be streamed to stderr on the parent.
+  child.stderr.pipe(process.stderr);
+
+  const instances = await getDetachedInstances();
```
Can we start the `getDetachedInstances` async function earlier (and `await` it only once we need it)?

jeffsmale90 replied:
yes! We now call `readdir` at the start of the function, and await it's response when we generate the instance name later on. Good call.

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,259 @@
+    if (shouldRemoveFile) removeDetachedInstanceFile(pid);
+  }
+
+  instances.sort((a, b) => b.startTime - a.startTime);
```
Most of the usages of `getDetachedInstances` don't need the return value to be sorted, and the name doesn't indicate it'll be sorted anyway. I think the sorting should be done by the caller that needs it sorted.

davidmurdoch replied:
Actually, this function may work really well as an async generator. You'd be able to implement: `findDetachedInstanceByName` without having to first iterate over all instances.

jeffsmale90 replied:
Good call - we not only call this function when calling `ganache instances list`, but sort at the consumer anyways.

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,259 @@
+ * @returns {Promise<DetachedInstance[]>} resolves with an array of instances
+ */
+export async function getDetachedInstances(): Promise<DetachedInstance[]> {
+  const files = readdirSync(dataPath);
```
Why do you use synchronous methods instead of the async counterparts?

jeffsmale90 replied:
We now async all the things!

davidmurdoch commented on src/packages/cli/src/process-name.ts:
```diff
@@ -0,0 +1,85 @@
+ * `<adjective>_<flavor>_<type>`, eg., `salted_caramel_ganache`.
+ */
+export default function createInstanceName() {
+  const name = [adjectives, flavors, kinds].map(pick).join("_");
```
This current code doesn't bother me, so feel free to ignore this, but the below suggestion avoids the creation of an array and two loops. :-p

```suggestion
  const name = `${pick(adjectives)}_${pick(flavors)}_${pick(kinds)}`;
```

jeffsmale90 replied:
Yeah this is probably better.

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,259 @@
+    version
+  };
+
+  const instanceFilename = `${dataPath}/${instance.pid}`;
```
I think we would be able to speed up some other areas of instance management if we stored the file with the `instanceName` as the file name instead of its `pid`. We wouldn't need to read all instances to get a list of names, as we could just check if a file already exists with the name we randomly picked, and when we need to manage (`stop`) an instance we can get it's file directly without having having to iterate over all files.

Is there a reason to use `pid`?

jeffsmale90 replied:
There _was_ - I think in a previous iteration, the child instance needed access to the instance file for some reason 🤷 

That's no longer the case - I've updated it to use the friendly name as the filename, which definitely improves finding the instance details by name. 👍

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,259 @@
+export function removeDetachedInstanceFile(pid: number): boolean {
+  const instanceFilename = `${dataPath}/${pid}`;
+  if (existsSync(instanceFilename)) {
+    rmSync(instanceFilename);
```
File system access is weird since time passes between the call to `existsSync` and `rmSync`, you'll still need to `try`/`catch` around `rmSync`.

Since you need to `try`/`catch` `rmSync`, there is no reason check `existsSync`.

In any case, if you are going to use `rmSync` you should make use of the useful options it gives you (like `force` and `retries`). And you should also probably use the async version, `rm`. If you don't want to use any of the `rm*` options, `unlink` is probably better, as it prevents the accidental deletion of directories -- which is always nice, even if it's not possible here (is it? :sweat_smile:).

jeffsmale90 replied:
I've try/catch'd the rmSync call - I will look into opportunities to improve performance through async fs access and tie the conversation to your question "Why do you use synchronous methods instead of the async counterparts?".

davidmurdoch reviewed (commented):
just another partial review

jeffsmale90 committed:
Move instance list sorting to the consumer that needs the list sorted

jeffsmale90 committed:
use detached instance friendly name for instance file (instead of pid)

jeffsmale90 committed:
pass original process.args to startDetachedInstance, calculate expected process 'cmd' value.

jeffsmale90 committed:
Fix 'cmd' argument for win32 platform

jeffsmale90 committed:
Add attribution for duration formatter

jeffsmale90 committed:
Change all detach mode file access to async

jeffsmale90 committed:
Adds warning messages if instances removed

jeffsmale90 committed:
Add PID to warning message.

davidmurdoch commented on src/packages/cli/tests/detach.test.ts:
```diff
@@ -0,0 +1 @@
+// todo: needs tests
```
You can use something like

```
describe("@ganache/cli", () => {
  describe(detach", () => {
    it("needs tests");
  });
});
```

which has adds it to the skipped tests in our tests runs; which to me, are just begging to be looked and fixed. You can also add a `// todo` comment if you'd like.

jeffsmale90 replied:
Good idea!

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,250 @@
+  } catch (err) {
+    return false;
+  } finally {
+    removeDetachedInstanceFile(instanceName);
```
should this be

```suggestion
    await removeDetachedInstanceFile(instanceName);
```
?

jeffsmale90 replied:
Good catch

jeffsmale90 committed:
Remove unnecessary detach.test.ts

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,250 @@
+  let instanceName: string;
+  do {
+    instanceName = createInstanceName();
+  } while (instanceNames.indexOf(instanceName) !== -1);
```
This is racey. Instead of reading all files in the directory then checking if the random instance name previously existed in the past, we could do _something_ like this completely untested code: 

```
while (true) {
  try {
   const instanceName = createInstanceName();
   const instanceFileName = getInstanceFileName(instanceName);
   details.instanceName = instanceName;
   await writeFile(instanceFilename, details, { flag: "wx" }); // wx means "write, but fail if the file exists"
   break;
  }
  catch(e) {
    if (e.code !== "EEXIST") throw e;
  }
}
```

jeffsmale90 replied:
Good find!

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,250 @@
+    }
+    reject(path);
+  }
+  resolve(path);
```
The directory should probably be created only at the moment in which we attempt to write a file to the directory. It doesn't need to exist for reads.

This has a nice benefit of not requiring `await dataPath` in these case when we just want to _read_ or delete from the directory, and it also doesn't cause a random filesystem call on all Ganache start ups (we don't need to proactively create the directory, so I don't think we should).

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,250 @@
+const START_ERROR = "An error ocurred spawning a detached instance of Ganache:";
+
+const dataPath = new Promise<string>(async (resolve, reject) => {
+  const path = envPaths(`Ganache/instances`).data;
```
@tenthirtyone Does this path play nicely with version-check?

jeffsmale90 replied:
I stole the path suffix from [persistent-cache.ts](https://github.com/trufflesuite/ganache/blob/develop/src/chains/ethereum/ethereum/src/forking/persistent-cache/persistent-cache.ts#L106)

davidmurdoch replied:
Ah, good call! @tenthirtyone, we might want to use capital `G` `ganache` in version-check :thinking:

tenthirtyone replied:
Ah here I thought you were referring to my comments. 

So I was all "Ah I'm tagged in chat for this PR I approved. I'll check it to finish on Friday" and here I find I am holding up stuff.

tenthirtyone replied:
@davidmurdoch @jeffsmale90 -> My late to the party post mortem here.

I have to make some assumptions about what plays nicely means.

If it means "This is separate from VC entirely, they will never clobber each other" - maybe, should be, probably? `Conf` has a default name for its file but we can also control that name.

If it means "We are going to combine these to reduce the number of disk reads" then...

w/r/t version-check we made the decision to use `Conf` because Truffle used conf and we wanted to support both using the same config at some point, I believe. So multi tenant Ganache/Truffle in the same file. We explicitly wanted to keep it separate from Ganache config. 

And version-check was never to be tightly coupled to Ganache because it was originally intended to be a separate lib altogether at some point. 

I have no preference either way but need the clarity on whether VC is still intended to standalone or not. 

If it does change I would still like to review version-check as is since it satisfies the original spec and treat the potential new config work as an issue for future work instead of moving the goal post and prolonging the existing review. 

We also talk about VC as being something that is not changed much after releasing it. 

And my two cents -> Our config system should support file-based configs and this should all go in there.

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,250 @@
+export async function getDetachedInstances(): Promise<DetachedInstance[]> {
+  const files = await readdir(await dataPath);
+  const instances: DetachedInstance[] = [];
+  const processes = await psList();
```
These two async calls can be parallelized.

davidmurdoch commented on src/packages/cli/src/cli.ts:
```diff
@@ -38,109 +65,182 @@ const isDocker =
+    // instance), so we need to notify that we are ready.
+    const isDetachedInstance = process.send !== undefined;
+    if (isDetachedInstance) {
+      notifyDetachedInstanceReady();
```
Why are we notifying at this point? The detached instance is _not_ ready at this point, as the server hasn't even been started yet. :thinking:

jeffsmale90 replied:
AFAICT the server _is_ started - this is in the callback passed into `server.listen`.

I've refactored this a touch - although I'm not a huge fan of the inline callback, I think this makes the logic flow a little more clear.

davidmurdoch replied:
Sorry, you are right. I totally misread the original code!

davidmurdoch reviewed (changes requested):
another partial review

jeffsmale90 committed:
improvements / refactor as per pr review.
- only create instances directory if we are writing
- check whether a name is taken by attempting to write to it
- correctly await removal of instance file when stopping an instance
- correct some wordsmithing

jeffsmale90 committed:
Don't check for existence of the folder each time - just catch the Error and create it.

jeffsmale90 committed:
Accidentally recreated the instance name twice

jeffsmale90 committed:
Move comment to relevant code

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+  version: string;
+};
+
+type CodedError = Error & { code: string };
```
I tihnk node's type for this is `NodeJS.ErrnoException`

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+type CodedError = Error & { code: string };
+
+const READY_MESSAGE = "ready";
+const START_ERROR = "An error ocurred spawning a detached instance of Ganache:";
```
```suggestion
const START_ERROR = "An error occurred spawning a detached instance of Ganache:";
```

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+
+async function getInstanceFilePath(instanceName: string): Promise<string> {
+  return path.join(await dataPath, instanceName);
+}
```
This shouldn't be async

davidmurdoch commented on src/packages/cli/src/cli.ts:
```diff
@@ -38,109 +65,181 @@ const isDocker =
-      rl.close();
-      handleSignal("SIGINT");
+    .catch(err => {
+      // the child process would have output it's error to stdout, so no need to
```
```suggestion
      // the child process would have output its error to stdout, so no need to
```

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+  let files: string[];
+  let processes: ProcessDescriptor[];
+  try {
+    [files, processes] = await Promise.all([readdir(await dataPath), psList()]);
```
```suggestion
    [files, processes] = await Promise.all([readdir(dataPath), psList()]);
```

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+    try {
+      // getDetachedInstanceByName() throws if the instance file is not found or
+      // cannot be parsed
+      const instance = await getDetachedInstanceByName(instanceName);
```
Is there a reason not to parallelize this loop?

jeffsmale90 replied:
Nope - I've updated this.

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+      const matchingProcess = processes.find(p => p.pid === instance.pid);
+      if (!matchingProcess) {
+        console.warn(
+          `Process not found; ${instanceName} has been removed.  (PID: ${instance.pid}).`
```
What about something more human sounding? Like:

```suggestion
          `Process with PID ${instance.pid} could not be found; removing ${instanceName} from recorded instances.`
```

More bikeshedding welcome!

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+        // if the cmd does not match the instance, the process has been killed,
+        // and another application has taken the pid
+        console.warn(
+          `Process information doesn't match; ${instanceName} has been removed. (PID: ${instance.pid}).`
```
And another more human sounding option:

```suggestion
          `Process with PID ${instance.pid} doesn't match; removing ${instanceName} from recorded instances.`
```

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+      shouldRemoveFile = true;
+    }
+
+    if (shouldRemoveFile) removeDetachedInstanceFile(instanceName);
```
Since `removeDetachedInstanceFile` is async, should this be `await`ed here? As is, if the deletion stalls for whatever reason, (/shrug), it will prevent the process from shutting down, and if it it throws it will _crash_ the process (though I know that `removeDetachedInstanceFile` wraps in a `try`/`catch`, so in this case it's impossible -- so throwing isn't really a concern)

jeffsmale90 replied:
yes!

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+ */
+async function getDetachedInstanceByName(
+  instanceName: string
+): Promise<DetachedInstance | undefined> {
```
This doesn't ever return `undefined`
```suggestion
): Promise<DetachedInstance> {
```

jeffsmale90 replied:
Ah true, it _used to_, but now it throws instead.

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+async function getDetachedInstanceByName(
+  instanceName: string
+): Promise<DetachedInstance | undefined> {
+  const filepath = await getInstanceFilePath(instanceName);
```
Since `getInstanceFilePath` doesn't need to be async we shouldn't await it (once you fix it).

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+  instanceName: string
+): Promise<DetachedInstance | undefined> {
+  const filepath = await getInstanceFilePath(instanceName);
+  const content = await readFile(filepath, { encoding: "utf8" });
```
You can `readFile` in as Buffer; `JSON.parse` knows how to stringify it directly. I seriously doubt there is a performance benefit to it, but since when we `writeFile` the file the encoding isn't specified I lean towards omitting it here, too (maybe the encoding should also be specified on the write side, just for explicitness-sake, even though the default is already `"utf8"`)

jeffsmale90 replied:
I've explicitly specified the encoding both when reading and writing.

In order to `JSON.parse` from buffer I would need to fix the typings, for dubious benefit. I'm open to dissenting opinions though :)

davidmurdoch commented on src/packages/ganache/npm-shrinkwrap.json:
```diff
@@ -487,9 +487,9 @@
-			"integrity": "sha512-asabaBSkEKosYKMITunzX177CXxQ4Q8BSSzMTKD+FefUhipQC70gfW5SiUDhYQ3vk8G+81HqQk7Fv9OXwwn9KA==",
+			"version": "8.8.0",
+			"resolved": "https://registry.npmjs.org/acorn/-/acorn-8.8.0.tgz",
+			"integrity": "sha512-QOxyigPVrpZ2GXT+PFyZTl6TtOFc5egxHIP9IlQ+RbupQuX4RkT/Bee4/kQuC02Xkzg84JcT7oLYtDIQxp+v7w==",
```
Since no `ganache` package deps changed lets omit this change from the PR.

davidmurdoch replied:
boop

davidmurdoch commented on src/packages/cli/package.json:
```diff
@@ -59,9 +59,12 @@
+    "nyc": "15.1.0"
   },
   "dependencies": {
+    "cli-table": "^0.3.1",
```
This needs to be pinned (and package-lock.json updated to match)

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+  try {
+    await rm(instanceFilename);
+    return true;
+  } catch (err) {}
```
```suggestion
  } catch {}
```

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+ * corresponding instance file will be removed.
+ *
+ * Note: This does not guarantee that the instance actually stops.
+ * @param  {string} instanceName
```
Just FYI, no need to remove them, but specifying the types in the JSDoc aren't needed, as they are redundant because we already defined the types in TypeScript.

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,261 @@
+): Promise<DetachedInstance> {
+  const [bin, module, ...args] = argv;
+  const childArgs = args.filter(
+    arg => arg !== "--detach" && arg !== "--😈" && arg !== "-D"
```
This fails with variants of `ganache --detach=1`

jeffsmale90 replied:
🙀
It's a shame that it isn't treated as a flag, because stripping out the detach argument becomes more complicated. I've implemented it with a regex check, and covered it in tests.

davidmurdoch replied:
At least you don't need to handle `--no-detach` (which _is_ a thing) :-p

jeffsmale90 committed:
Improvements to detach re PR

jeffsmale90 committed:
Revert package-lock.json

jeffsmale90 committed:
_Actually_ fix package-lock.json this time

jeffsmale90 committed:
Remove nyc dev-dependency from cli/package.json - correct, but irrelevant for this PR

jeffsmale90 committed:
Strip the detach argument correctly before passing arguments to child process

davidmurdoch commented on src/packages/cli/src/cli.ts:
```diff
@@ -28,6 +39,22 @@ const logAndForceExit = (messages: any[], exitCode = 0) => {
+    .filter(val => val[1] !== 0)
+    .map(([key, val]) => `${val} ${key}${val !== 1 ? "s" : ""}`)
+    .join(", ");
+};
```
If seconds is computed as `0` this returns "". Maybe that's okay?

jeffsmale90 replied:
I've added a fallback as "Just started" in the case that someone starts an instance and immediately lists them, as well as some test cases for those.

davidmurdoch commented on src/packages/cli/package.json:
```diff
@@ -70,6 +70,8 @@
     "ethereumjs-util": "7.1.5",
-    "marked-terminal": "4.1.0"
+    "marked-terminal": "4.1.0",
+    "cli-table": "^0.3.1",
```
![image](https://user-images.githubusercontent.com/187813/202547862-84dc1649-4914-4599-873d-6370fc71d9a5.png)

jeffsmale90 replied:
oops :/

davidmurdoch commented on src/packages/ganache/README.md:
```diff
@@ -476,7 +476,43 @@ Server:
+
 
+$ ganache instances stop salted_caramel_ganache
+Process stopped
```
Can you add the `With the following command, you can start Ganache, run your tests, and stop Ganache when you are finished.` example to the README?

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,280 @@
+        followingArg !== undefined && followingArg[0] !== "-";
+
+      strippedArgs.splice(i, hasTrailingValue ? 2 : 1);
+      return strippedArgs;
```
This is a fun one: `ganache --detach --detach` :-p. Unfortunately we can't short circuit here because the flag could be present in the args multiple times.

jeffsmale90 replied:
Added cases for a couple different permutations of duplicate arguments. But, again, I wish this were simpler!

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,280 @@
+          // a very uncommon case. Catching the exception and subsequently
+          // creating the directory is faster in the majority of cases.
+          await mkdir(dataPath, { recursive: true });
+          break;
```
should this be `continue;`?

jeffsmale90 replied:
It _could_ be - the same could be said of each of the breaks within this switch statement - but I think break is a little simpler (at least for me) to consume - it's the last statement within the scope of the loop, so 🤷 - happy to change it thought.

davidmurdoch replied:
Er, no. i'm dumb. I was thinking this with the while's `break;`

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,280 @@
+
+  const flavor = instanceInfo.flavor;
+  const { host, port } = instanceInfo.server;
+  const cmd =
+    process.platform === "win32"
+      ? path.basename(process.execPath)
+      : [process.execPath, ...process.execArgv, module, ...childArgs].join(" ");
```
On a scale of 1 to Ethereum Deposit Contract, how confident are you on this?

jeffsmale90 replied:
I'm confident, under the assumption that `ps-list` has a consistent output on non-windows implementations. This is based on the output of the *nix `ps` command, which states in it's docs "Every operating system's version of ps is slightly different".

The implementation of `ps-list` clearly has had some effort made to normalise different implementations https://github.com/sindresorhus/ps-list/blob/0118e5d53b5d9af95b60983c9c67adb2ab742fbb/index.js#L139m so I think if there's going to be a problem, then it is likely going to be an obscure distribution.

jeffsmale90 replied:
I've tested on windows, Ubuntu and MacOS. I doubt we'd get much value by hunting out additional distros as it would need to be fairly obscure.

davidmurdoch replied:
That's probably good enough! @cds-amal might have something obscure on hand.

jeffsmale90 committed:
Add tests for duration formatting - move implementation into detach.ts

jeffsmale90 committed:
Rename formatDuration to formatUptime, add --detach example to README, handle --detach arg duplication

jeffsmale90 committed:
Revert shrinkwrap :/

jeffsmale90 committed:
Pin cli-table dependency to latest release

jeffsmale90 committed:
Fix failing test

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,298 @@
+        // if the cmd does not match the instance, the process has been killed,
+        // and another application has taken the pid
+        console.warn(
+          `Process with PID ${instance.pid} doesn't match ${instanceName}; removing${instanceName} from recorded instances.`
```
```suggestion
          `Process with PID ${instance.pid} doesn't match ${instanceName}; removing ${instanceName} from recorded instances.`
```

davidmurdoch commented on src/packages/cli/tests/detach.test.ts:
```diff
@@ -0,0 +1,102 @@
+        [60000, "1 minute"],
+        [62000, "1 minute, 2 seconds"],
+        [1000000, "16 minutes, 40 seconds"],
+        [-1000, "1 second"],
```
I didn't think anything of the `Math.abs` check, but now that I'm reviewing the tests I'm wondering why it was necessary, but more importantly, why "1 second" would be the desired output for an instances started 1 second in the future.

jeffsmale90 replied:
You are right, that's _not_ the desired result. I could argue, if the function was `formatDurationBetween(a, b)`, maybe `[a, b]` is not in order, but 🤷 

We still need to handle -ve values, so I've added a prefix of `In `- to all negative values. That means we get the likes of:

`should format an input of -171906000 as "In 1 day, 23 hours, 45 minutes, 6 seconds"`

davidmurdoch commented on src/packages/ganache/README.md:
````diff
@@ -476,7 +476,55 @@ Server:
+GANACHE=$(ganache --detach) && npm run test; ganache instances stop $GANACHE
+```
+
+Or if you are running Powershell on Windows, you can do:
````
```suggestion
Or if you are running PowerShell on Windows, you can do:
```

davidmurdoch commented on src/packages/ganache/README.md:
````diff
@@ -476,7 +476,55 @@ Server:
+Ganache can be started as a background instance via the CLI by providing the following argument (along with any combination
+of the Ganache startup arguments above):
+
+```
````
```suggestion
```console
```

davidmurdoch replied:
Note: i'm not 100% sure when to use `console` and when to use `bash` for this stuff.

jeffsmale90 replied:
So I had a look around, and take this with a grain of salt, because I don't have anything authoritative,  but:

https://github.com/github/linguist/blob/3c3b037910006fc2f1a9bb34b2c4e9cde062206c/lib/linguist/languages.yml shows the languages that github knows, and the various aliases.

- `bash` maps to `Shell`
- `console` maps to `ShellSession`

I believe that the difference is that a `Shell` block should be a shell _script_, where as a `ShellSession` should be a series of commands and responses from a session.

Given all of these examples are in the context of a _session_, I've used `console`. 

While I'm here, I standardised the console blocks, as some are missing the `$` prompt, which is somewhat confusing (at the very least, inconsistent!)

davidmurdoch commented on src/packages/ganache/README.md:
````diff
@@ -476,7 +476,55 @@ Server:
+receive requests. A friendly name will be returned to STDOUT which can then be used to interact with the instance via
+the `ganache instances` command with the following arguments:
+
+```
````
```suggestion
```console
```

davidmurdoch commented on src/packages/ganache/README.md:
````diff
@@ -476,7 +476,55 @@ Server:
+
+E.g., start Ganache on port 8544, with a block time of 10 seconds, and then stop the instance.
+
+```
````
```suggestion
```console
```

davidmurdoch commented on src/packages/ganache/README.md:
````diff
@@ -476,7 +476,55 @@ Server:
+
+With the following command, you can start Ganache, run your tests, and stop Ganache when you are finished.
+
+```
````
```suggestion
```bash
```

jeffsmale90 replied:
👍, except I used `console` as per my justification above.

davidmurdoch commented on src/packages/ganache/README.md:
````diff
@@ -476,7 +476,55 @@ Server:
+Or if you are running Powershell on Windows, you can do:
+
+```
+$GANACHE=ganache --detach; npm run test; ganache instances stop $GANACHE; Remove-Variable GANACHE
````
```suggestion
```PowerShell
GANACHE=ganache --detach; npm run test; ganache instances stop $GANACHE; Remove-Variable GANACHE
```

davidmurdoch commented on src/packages/ganache/README.md:
```diff
@@ -476,7 +476,55 @@ Server:
+
+### Detached Instances
+
+Ganache can be started as a background instance via the CLI by providing the following argument (along with any combination
```
```suggestion
Ganache can be started as a background instance via the CLI by providing the following argument (along with any valid combination
```

davidmurdoch commented on src/packages/cli/tests/detach.test.ts:
```diff
@@ -0,0 +1,102 @@
+      durations.forEach(duration => {
+        const [ms, formatted] = duration;
+        it(`should format an input of ${ms} as "${formatted}"`, () => {
+          const result = formatUptime(ms as number);
```
If you do `const durations: [number, string][]` above you won't need to cast here.

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,298 @@
+}
+
+const detachArgRegex = /^-(?:D|-detach|-😈)$|=/;
+export function stripDetachArg(args: string[]): string[] {
```
I'm so so sorry to find another one ... but this misses short args groups, like `-fdD` (equiv to `--fork --deterministic --detach`). https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html

Maybe we can just add `--no-detach` at the end of the args.

jeffsmale90 replied:
As discussed: 😩

jeffsmale90 replied:
I'm not _super_ happy with the approach, but it works, so 🤷 better than one that doesn't.

I've also added a test case that asserts that the args parser ignores any detach arguments, so long as it's proceeded by a `--no-detach` argument (which is, admittedly, a little tenuous).

jeffsmale90 committed:
Improve duration formatting, improve readme

jeffsmale90 committed:
Annoyingly, args parsing is complicated. Don't try to strip out the detach flags, rather append  to the child args, cancelling out all preceeding flags.

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,289 @@
+    try {
+      // getDetachedInstanceByName() throws if the instance file is not found or
+      // cannot be parsed
+      const instance = await getDetachedInstanceByName(instanceName);
```
Wonder if we need to watch out for invalid files, or if the `try`/`catch` here is enough.

I know that on MacOS if a user navigates to the directory (using Finder) the OS will drop a `.DS_Store` folder there (it stores computed metadata/settings for the directory). We might want to explicitly ignore this directory, or maybe put an extension on the instance name files (i.e., `.json`) so that we can filter for only files that could possibly be relevant (like `if (instanceName.endsWith(".json") ...`).

jeffsmale90 replied:
I don't think we _need_ to (invalid files are removed), but I do like the idea of a .json suffix. Originally these files weren't json, but as their complexity grew, I decided it was best.

davidmurdoch replied:
I guess the question was really 'Should we delete files that aren't "ours"?' :thinking:

davidmurdoch replied:
I normally wouldn't be concerned.... but since just using Finder to _read_ directories in turns _writes_ to those directories I'm not sure.

jeffsmale90 replied:
Agreed in person that we _should_ delete these files. In the latest revision, I've improved the error messaging around invalid files / removing unexpected stuff from the instances directory.

davidmurdoch commented on src/packages/cli/src/cli.ts:
```diff
@@ -38,109 +50,181 @@ const isDocker =
+        ]);
+      }
+
+      console.log(table.toString());
```
Made an issue for the rendering issue here: https://github.com/trufflesuite/ganache/issues/3949

jeffsmale90 committed:
Instance files now have .json extension. Remove any files / folders from instances directory that aren't valid json files. Rename 'poppy-seed' to 'poppyseed' because hyphens in the instance name are a pain!

jeffsmale90 committed:
Remove unused import

jeffsmale90 committed:
Handle errors betterer

jeffsmale90 commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,324 @@
+        // cannot be parsed
+        instance = await getDetachedInstanceByName(instanceName);
+      } catch (err: any) {
+        failureReason = err.message;
```
This is likely going to be a JSON parsing error - which is fairly informative, or a "file not found" or "access denied" sort of a thing.

jeffsmale90 commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,324 @@
+  await Promise.all(loadingInstancesInfos);
+
+  if (someInstancesFailed) {
+    console.warn(
```
I do this at the end, because if there's multiple errors, we only want to say this once.

davidmurdoch commented on src/packages/cli/src/detach.ts:
```diff
@@ -0,0 +1,324 @@
+const READY_MESSAGE = "ready";
+const START_ERROR =
+  "An error occurred spawning a detached instance of Ganache:";
+const dataPath = envPaths(`Ganache/instances`).data;
```
We don't uses a suffix for the persistent cache `db`, let's also drop it here:

```suggestion
const dataPath = envPaths("Ganache/instances", {
  suffix: ""
}).data;
```

jeffsmale90 replied:
My instances!

davidmurdoch commented on src/packages/cli/tests/detach.test.ts:
```diff
@@ -0,0 +1,34 @@
+      durations.forEach(duration => {
+        const [ms, formatted] = duration;
+        it(`should format an input of ${ms} as "${formatted}"`, () => {
+          const result = formatUptime(ms as number);
```
```suggestion
          const result = formatUptime(ms);
```

jeffsmale90 committed:
Shorten duration formatting, combine host and port, right align PID in instances table. Revert changes to standardise prompts in README

jeffsmale90 committed:
RH align Uptime

jeffsmale90 committed:
remove unnecessary type assertion, remove datapath suffix

cds-amal commented on src/packages/cli/src/detach.ts:
```diff
@@ -2,9 +2,9 @@ import { fork } from "child_process";
+import { Dirent, promises as fsPromises } from "fs";
 // this awkward import is required to support node 12
-const { readFile, rm, mkdir, readdir, writeFile } = fsPromises;
+const { readFile, rm, mkdir, readdir, writeFile, unlink } = fsPromises;
```
Is Node 12 still an issue for ganache?

davidmurdoch replied:
Until the merge work is in, yes.

cds-amal commented on src/packages/cli/src/detach.ts:
```diff
@@ -219,52 +221,70 @@ export async function getDetachedInstances(): Promise<DetachedInstance[]> {
+      if (dirEntry.isDirectory()) {
+        const reason = `"${filename}" is a directory`;
+        try {
+          await rm(fullPath, { recursive: true });
```
Can this be used with the `force: true` option, so there's no need for `unlink` as well? It seems `rm` should work on both files and folders... but oops, is it available in [node 12](https://nodejs.org/dist/latest-v18.x/docs/api/fs.html#fsrmpath-options-callback)?

davidmurdoch replied:
Ah, great catch! This party doesn't have tests yet.

jeffsmale90 replied:
I've created https://github.com/trufflesuite/ganache/issues/3965 to track this improvement.

Unfortunately rm wasn't added until 14.14 :(

I've updated the call to rm() to use `rmDir` to be nice to node 12 (even though we handled that case reasonably well).

jeffsmale90 replied:
I'm working on integration tests for this here: #3735 - they work, and they have reasonable coverage, but they are not especially pretty yet.

cds-amal reviewed (commented):
Drive by feedback. I'm excited to get this in my hands. OR, get my hands on this. wheeeeee

jeffsmale90 committed:
Use unlink instead of rm, because node 12

jeffsmale90 committed:
Support ndoe 12 correctly for removal of files and directories

davidmurdoch reviewed (approved):
SHIP IT!

jeffsmale90 committed:
Make PowerShell example consistent with non-Windows example
